Feature/create routes lists

Esodey opened the pull request:
- [x] Route to get all of a users lists

- [x] Route to get a single list

- [x] Route to post a list

~~- [ ] Route to update a list~~

~~- [ ] Route to delete a list~~

Esodey committed:
"Creat post route and test for lists"

Esodey committed:
Merge branch 'dev' into feature/Create-routes-lists

Esodey committed:
"Create get, post, put and delete routes for lists"

aeciorc commented on server/app/controllers/lists.js:
```diff
@@ -0,0 +1,134 @@
+ */
+// eslint-disable-next-line no-unused-vars
+const postList = asyncHandler(async (req, res, next) => {
+  const { name, coverUrl, user, products } = req.body;
```
I think here and below, you would take an actual image file instead of a url, which you would then use the s3 service (I think Simon is working on it) to upload it and get that url

aeciorc commented on server/app/models/List.js:
```diff
@@ -6,32 +6,38 @@ const options = {
+    },
+    coverUrl: {
+      type: String,
+      required: [true, 'Please add a cover URL'],
```
Maybe this shouldn't be required. We could have a default image for an empty list, and if there's no coverUrl, the frontend could display that image or the image of the first product on that list. What do you guys think, @Esodey and @vic2019 ?

Esodey replied:
That makes sense to me

maismin commented on server/app/controllers/lists.js:
```diff
@@ -0,0 +1,134 @@
+});
+
+/**
+ * @api {update} /api/v1/lists/:listId
```
The HTTP method should be inside the curly braces, so replace update with put.

Esodey committed:
"Change api from update to put"

maismin commented on server/app/controllers/lists.js:
```diff
@@ -0,0 +1,134 @@
+  if (!lists) {
+    return next(
+      new ErrorResponse(
+        `No lists with the user id of ${req.params.userId}`,
```
You've destructured the `userId` from `req.params`, so `req.params.userId` can be replaced with `userId`

maismin commented on server/app/controllers/lists.js:
```diff
@@ -0,0 +1,134 @@
+ * @apiGroup lists
+ * @apiPermission protected
+ *
+ * @apiDescription Creates user in database
```
Fix comment

maismin commented on server/app/controllers/lists.js:
```diff
@@ -0,0 +1,134 @@
+
+  await list.remove();
+  return res.status(200).json({ success: true, data: list });
+});
```
Since the List document has references to Products and vice versa, for each Product in the List you have to delete those references to the List. Moreover, if the Product's list property is empty (it's not in any list) then the product should be deleted.

Although @Esodey and I spoke about including a delete route for shopping lists, the sketch does not include any way to delete a list. Perhaps we should exclude the delete route and include it later. What do you think @aeciorc ?

maismin commented on server/app/controllers/lists.js:
```diff
@@ -0,0 +1,134 @@
+    success: true,
+    list: list.toJSON(),
+  });
+});
```
After reviewing the sketch, I don't think we need this route since the only time the list is updated is when a new product is added to the list, in which case a `POST` to `/api/v1/lists/:listId/products` is made instead.

Esodey committed:
"Cut required for coverUrl. Fix use of userId variable"

maismin committed:
Merge branch 'dev' into feature/Create-routes-lists

maismin committed:
Add more files for seeding database

- Replace passwords with hash values with actual values in user.json.
Passing the hashed passwords causes them to be hashed again when
populating the database.

- Add lists.json to populate the lists collections

- Change products in products.json to be added to the lists in
lists.json

maismin committed:
Fix API routes for lists

- Remove update and delete routes as they are not needed for the MVP.

- Add authentication to GET routes.

- Fix POST route to upload cover images to S3.

- Add followers field for List model to incorporate followers.

- Rename user field in List model to creator.

Esodey commented on server/app/_data/lists.json:
```diff
@@ -0,0 +1,32 @@
+[
```
Finally some real mock data

Esodey commented on server/app/_data/seeder.js:
```diff
@@ -9,6 +9,7 @@ dotenv.config({
 });
 
 // Load models
+const List = require('../models/List');
```
Very useful going forward. Nice to see a well designed seeder

Esodey commented on server/app/controllers/lists.js:
```diff
@@ -0,0 +1,129 @@
+  const { userId } = req.params;
+  const user = req.user.toJSON();
+
+  if (user.id !== userId) {
```
Nice additon

Esodey commented on server/app/controllers/lists.js:
```diff
@@ -0,0 +1,129 @@
+
+  const lists = await List.find({ followers: userId })
+    .select('-followers')
+    .populate({
```
Good catch on this. Good to know .populate is a thing

Esodey commented on server/app/controllers/lists.js:
```diff
@@ -0,0 +1,129 @@
+  const user = req.user.toJSON();
+
+  if (!req.files) {
+    list = await List.create({ name, creator: user.id, followers: [user.id] });
```
good check

Esodey commented on server/app/models/List.js:
```diff
@@ -6,32 +6,45 @@ const options = {
+      ref: 'User',
+      required: true,
+    },
+    followers: {
```
should def bring this up on monday

Esodey reviewed (commented):
All looks great. Nice work
